src/fprime_extras/lint/rules/rule_valid_port_types.py

```diff
@@ -0,0 +1,162 @@
+import logging
+
+from ...core.hooks import HookRegistry as hooks
+from .base import AbstractRule
+from .base import LintSeverity
+from .base import Notification
+from .registry import RuleListRegistry
+from ...core.util import find_fprime
+
+from pathlib import Path
+import os, sys
+import lxml.etree as etree
+from lxml.etree import XMLSyntaxError
+
+log = logging.getLogger(__name__)
+slug = 'invalid-port-type'
+# TODO: import following list from model instead of defining it here.
+port_kinds = ['output', 'sync_input', 'async_input', 'guarded_input']
+
+
+@RuleListRegistry.register()
+class RuleDanglingPortConnections(AbstractRule):
+    """Check if input XML file can be parsed."""
+
+    def __init__(self):
+        super().__init__()
+
+    @classmethod
+    def register_hook_callbacks(cls):
+        hooks.register('fprime_extras/lint/document_model_rules', cls())
+
+    def __call__(self, config, extras_file, doc_model):
+        notifications = []
+        fprime_path = config.fprime_root
+        if fprime_path is None or not os.path.exists(fprime_path):
+            log.info("Received input root of {} which is invalid. Searching for Fprime root instead".format(fprime_path))
+            if fprime_path is not None:
+                msg = "{} does not exist. Searching for Fprime root as parent of component.".format(fprime_path)
+                note = Notification(filename=extras_file.full_filename,
+                                    rule_slug=slug,
+                                    description=msg,
+                                    line_number=0,
+                                    column=0,
+                                    severity=LintSeverity.WARNING
+                                    )
+                notifications.append(note)
+            fprime_path = find_fprime(Path(extras_file.full_filename))
+        if fprime_path is None:
+            descr = "Cannot check port types. Missing valid Fprime path."
+            descr += " Please execute command from Fprime subdirectory or pass Fprime path as arg."
+            note = Notification(filename=extras_file.full_filename,
+                                rule_slug=slug,
+                                description=descr,
+                                line_number=0,
+                                column=0,
+                                severity=LintSeverity.WARNING
+                                )
```

Comment:
This lint warning is OK, because the lack of availability of this path kills the ability to do an important check, but doesn't prevent the linter from checking other things.

Reply:
BTW: this is effectively a different check, so it should be a different rule slug.  You can see how multiple slugs are used in other rules.  The real issue with slugs is that they define the filter granularity for ignoring errors, so if it would be preferable to ignore a warning about missing the F Prime root, but not ignore the warning about bad port kind specifications; then we need them to have different slugs.

Reply:
I changed this slug. I wasn't sure how to incorperate the slugs in a way like the fplint rule, so I just used 3 string variables for the 3 different slugs

Reply:
looks good